cudaextender/src/ungapped_xdrop_kernels.cu

```diff
@@ -0,0 +1,493 @@
+            {
+
+                if (max_pos > prev_max_pos[warp_id])
+                    new_max_found[warp_id] = true;
+                else
+                    new_max_found[warp_id] = false;
+
+                if (xdrop_done)
+                {
+                    total_score[warp_id] += max_thread_score;
+                    xdrop_found[warp_id] = true;
+                    left_extent[warp_id] = max_pos;
+                    extent[warp_id] += left_extent[warp_id];
+                    prev_max_pos[warp_id] = max_pos;
+                    tile[warp_id]         = max_pos;
+                }
+                else if (ref_loc[warp_id] < pos_offset || query_loc[warp_id] < pos_offset)
+                {
+                    total_score[warp_id] += max_thread_score;
+                    edge_found[warp_id]  = true;
+                    left_extent[warp_id] = max_pos;
+                    extent[warp_id] += left_extent[warp_id];
+                    prev_max_pos[warp_id] = max_pos;
+                    tile[warp_id]         = max_pos;
+                }
+                else
+                {
+                    prev_score[warp_id]     = thread_score;
+                    prev_max_score[warp_id] = max_thread_score;
+                    prev_max_pos[warp_id]   = max_pos;
+                    tile[warp_id] += warp_size;
+                }
+            }
+            __syncwarp();
+
+            if (new_max_found[warp_id])
+            {
+                for (int32_t i = 0; i < 4; i++)
+                {
+                    count[i]     = count[i] + count_del[i];
+                    count_del[i] = 0;
+                }
+            }
+            __syncwarp();
+
+            if (r_chr == q_chr)
+            {
+                if (pos_offset <= prev_max_pos[warp_id])
+                {
+                    count[r_chr] += 1;
+                }
+                else
+                {
+                    count_del[r_chr] += 1;
+                }
+            }
+            __syncwarp();
+        }
+
+        //////////////////////////////////////////////////////////////////
+
+        if (total_score[warp_id] >= score_threshold && total_score[warp_id] <= 3 * score_threshold && !no_entropy)
+        {
+            for (int32_t i = 0; i < 4; i++)
+            {
+#pragma unroll
+                for (int32_t offset = 1; offset < warp_size; offset = offset << 1)
+                {
+                    count[i] += __shfl_up_sync(0xFFFFFFFF, count[i], offset);
+                }
+            }
+            __syncwarp();
+
+            if (lane_id == warp_size - 1 && ((count[0] + count[1] + count[2] + count[3]) >= 20))
+            {
+
+                entropy[warp_id] = 0.f;
+                for (int32_t i = 0; i < 4; i++)
+                {
+                    entropy[warp_id] += ((double)count[i]) / ((double)(extent[warp_id] + 1)) * ((count[i] != 0) ? log(((double)count[i]) / ((double)(extent[warp_id] + 1))) : 0.f);
```

Comment:
I think this could use some local variables to ease understanding.

Reply:
Added a local variable for probability and made the loop more efficient.

Reply:
:+1: 
Minor optimization: you could save a bit of shared memory traffic by introducing a local `new_entropy` variable and only do a single write to the array `entroy[warp_id]`.

Reply:
Done. Renamed the intermediate var to `entropy_ln` and `log_4` to `ln_4`.